### DISCUSSION TITLE
Should isSerializableProps() allow non-plain objects with a toJSON() method?

### DISCUSSION THREAD

**Discussion by sa3dany** [vercel/next.js#28458] [Ideas]
JavaScript's built-in `JSON.stringify()` checks for the existence of a [`toJSON` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#tojson_behavior) when serializing the passed in object (it will also do this for nested objects). It then uses the returned object from `toJSON()` when serializing instead of using the original object.

Right now in Next.js, [JSON is used to pass server-side props to the front end from `getServerSideProps()`, etc...](https://github.com/vercel/next.js/blob/51559f5c645c6ad8ba1bf2da74bac284015f887e/packages/next/server/render.tsx#L965). 

During development and when running `next build`, Next.js tests props returned from `getServerSideProps()`, etc... using [`isSerializableProps()`](https://github.com/vercel/next.js/blob/51559f5c645c6ad8ba1bf2da74bac284015f887e/packages/next/lib/is-serializable-props.ts#L16) which makes sure all the values passed can be serialized to JSON (i.e. objects and arrays can only contain primitives, excluding `undefined`). However, `isSerializableProps()` does not consider the existence of `toJSON()` when validating props.

So the question is: **Does it make sense for `isSerializableProps()` to use `toJSON()` when validating props?**

## Workaround
```js
// For example the return value inside `getServerSideProps()`
return {
  props: {
    customType: customType.toJSON(),
  }
}
```

**Comment by GabenGar** [comment]
But how are you going to recreate the class instance off the json string in the page props and statically analyze all of that?

**Reply by sa3dany** [reply]
That's not my point. If the user want to recreate a class instance that's their responsibility and he can do that from page props. This is just about `toJSON()` on objects being called by `JSON.stringify()` and `isSerializableProps()` does not take this into consideration.

**Reply by GabenGar** [reply]
Actually it is related to the point. As of now the props returned by the server function and passed to the page component are governed by the same interface. Allowing random classes in there means the underlying type system also has to be rewritten with runtime checks on top. So it's not just a user's problem.
